### PR TITLE
Slashing operations pool

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

# Description

Adds a slashing operation pool for both attester and proposer slashings.

### CODE REVIEW

File: beacon-chain/operations/slashings/service.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+// Pool implements a struct to maintain pending and recently included voluntary exits. This pool
+// is used by proposers to insert into new blocks.
+type Pool struct {
+	lock                    sync.RWMutex
+	pendingProposerSlashing []*ethpb.ProposerSlashing
+	pendingAttesterSlashing []*PendingAttesterSlashing
+	included                map[uint64]bool
+}
+
+// PendingAttesterSlashing represents an attester slashing in the operation pool.
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go]: Can you define this on another file? `types.go` ?

File: beacon-chain/operations/slashings/service.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+// Allows for easy binary searching of included validator indexes.
+type PendingAttesterSlashing struct {
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go]: How come we dont do the same for proposer?

File: beacon-chain/operations/slashings/service.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	attesterSlashing *ethpb.AttesterSlashing
+	validatorToSlash uint64
+}
+
+// NewPool accepts a head fetcher (for reading the validator set) and returns an initialized
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go]: where is the head fetcher as an input?

File: beacon-chain/operations/slashings/service.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+// slashed validator pool.
+func NewPool() *Pool {
+	return &Pool{
+		pendingProposerSlashing: make([]*ethpb.ProposerSlashing, 0),
+		pendingAttesterSlashing: make([]*PendingAttesterSlashing, 0),
+		included:                make(map[uint64]bool),
+	}
+}
+
+// PendingAttesterSlashings returns exits that are ready for inclusion at the given slot. This method will not
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go]: It doesn't return the exit

File: beacon-chain/operations/slashings/service.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+	for i, slashing := range p.pendingProposerSlashing {
+		if i >= int(params.BeaconConfig().MaxProposerSlashings) {
+			break
+		}
+		pending = append(pending, slashing)
+	}
+	return pending
+}
+
+// InsertAttesterSlashing into the pool. This method is a no-op if the pending exit already exists,
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go]: not `pending exit`, please edit that

File: beacon-chain/operations/slashings/service.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/operations/slashings/service.go]: +1

File: beacon-chain/operations/slashings/service.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+// PendingAttesterSlashings returns exits that are ready for inclusion at the given slot. This method will not
+// return more than the block enforced MaxAttesterSlashings.
+func (p *Pool) PendingAttesterSlashings() []*ethpb.AttesterSlashing {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	var included []uint64
+	pending := make([]*ethpb.AttesterSlashing, 0)
```
Comment (nisdas) [inline beacon-chain/operations/slashings/service.go]: set the capacity to the max limit

File: beacon-chain/operations/slashings/service.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+	}
+}
+
+// PendingAttesterSlashings returns exits that are ready for inclusion at the given slot. This method will not
+// return more than the block enforced MaxAttesterSlashings.
+func (p *Pool) PendingAttesterSlashings() []*ethpb.AttesterSlashing {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	var included []uint64
```
Comment (nisdas) [inline beacon-chain/operations/slashings/service.go]: same here set a capacity for this

File: beacon-chain/operations/slashings/service.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+	return pending
+}
+
+// PendingProposerSlashings returns proposer slashings that are ready for inclusion at the given slot.
+// This method will not return more than the block enforced MaxProposerSlashings.
+func (p *Pool) PendingProposerSlashings() []*ethpb.ProposerSlashing {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	pending := make([]*ethpb.ProposerSlashing, 0)
```
Comment (nisdas) [inline beacon-chain/operations/slashings/service.go]: same here,set the capacities before hand

File: beacon-chain/operations/slashings/service.go (hunk 2 above)
Comment (0xKiwi) [reply beacon-chain/operations/slashings/service.go]: Every proposal slashing only has one validator index to slash, not multiple like a attestation slashing can have. So we wrap it in order to easily sort and binary search them.

File: beacon-chain/operations/slashings/service.go (hunk 5 above)
Comment (0xKiwi) [reply beacon-chain/operations/slashings/service.go]: Apologies, it was initially copy pasted from exits.

File: beacon-chain/operations/slashings/service.go [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	var included []uint64
+	pending := make([]*ethpb.AttesterSlashing, 0)
+	for i, slashing := range p.pendingAttesterSlashing {
+		if i >= int(params.BeaconConfig().MaxAttesterSlashings) {
+			break
+		}
+		if sliceutil.IsInUint64(slashing.validatorToSlash, included) {
```
[the remark this answers is not in the document]
Comment (0xKiwi) [reply beacon-chain/operations/slashings/service.go]: Good catch, thank you

File: beacon-chain/operations/slashings/service.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		}
+		validator, err := state.ValidatorAtIndexReadOnly(val)
+		if err != nil {
+			return err
+		}
+		// Has the validator been exited already?
+		if validator.ExitEpoch() < helpers.CurrentEpoch(state) {
+			continue
+		}
+		//Has the validator been slashed already?
```
Comment (rauljordan) [inline beacon-chain/operations/slashings/service.go]:
```suggestion
		// Checking if the validator has already been slashed.
```

File: beacon-chain/operations/slashings/service.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// InsertProposerSlashing into the pool. This method is a no-op if the pending slashing already exists,
+// has been included recently, the validator is already exited, or the validator was already slashed.
+func (p *Pool) InsertProposerSlashing(state *beaconstate.BeaconState, slashing *ethpb.ProposerSlashing) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	idx := slashing.ProposerIndex
+
+	// Has this validator index been included recently?
```
Comment (rauljordan) [inline beacon-chain/operations/slashings/service.go]:
```suggestion
	// Check if this validator index been included recently.
```

File: beacon-chain/operations/slashings/service.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+	if p.included[idx] {
+		return nil
+	}
+
+	validator, err := state.ValidatorAtIndexReadOnly(idx)
+	if err != nil {
+		return err
+	}
+	// Has the validator been exited already?
```
Comment (rauljordan) [inline beacon-chain/operations/slashings/service.go]:
```suggestion
	// Check if the validator has already exited.
```

File: beacon-chain/operations/slashings/service.go [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+	if validator.ExitEpoch() < helpers.CurrentEpoch(state) {
+		return nil
+	}
+
+	// Has the validator been slashed already?
```
Comment (rauljordan) [inline beacon-chain/operations/slashings/service.go]:
```suggestion
	// Check if the validator has already been slashed.
```

File: beacon-chain/operations/slashings/service.go [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+	slashedValidators := state.Slashings()
+	found := sort.Search(len(slashedValidators), func(i int) bool {
```
Comment (rauljordan) [inline beacon-chain/operations/slashings/service.go]: I see duplicated code here and in the func above, could we refactor for deduplication or use a helper?

Comment (rauljordan) [reply beacon-chain/operations/slashings/service.go]: Resolved

**Comment by rauljordan** [comment]
Resolved all comments PTAL @nisdas

**Pull request merged**